### PR TITLE
Finish the Java Api

### PR DESCRIPTION
**Pull request by mudathirmahgoub**
This PR finishes the implementation of the Java API.
It also includes all java files in the build along with their unit tests.

### CODE REVIEW

File: src/api/java/cvc5/Solver.java [hunk 1]
```diff
@@ -26,36 +27,2502 @@ public long getPointer()
...
+    Utils.validateUnsigned(exp, "exp");
+    Utils.validateUnsigned(sig, "sig");
+    long termPointer = mkNegInf(pointer, exp, sig);
+    return new Term(this, termPointer);
+  }
+
+  private native long mkNegInf(long pointer, int exp, int sig);
+  /**
+   * Create a not-a-number (NaN) floating-point constant. Requires CVC4 to
+   * be compiled with SymFPU support.
```
Comment (4tXJ7f) [inline src/api/java/cvc5/Solver.java:1065]:
```suggestion
   * Create a not-a-number (NaN) floating-point constant.
```

File: src/api/java/cvc5/Solver.java [hunk 2]
```diff
@@ -26,36 +27,2502 @@ public long getPointer()
...
+    Utils.validateUnsigned(sig, "sig");
+    long termPointer = mkNaN(pointer, exp, sig);
+    return new Term(this, termPointer);
+  }
+
+  private native long mkNaN(long pointer, int exp, int sig);
+
+  /**
+   * Create a positive zero (+0.0) floating-point constant. Requires CVC4 to
+   * be compiled with SymFPU support.
```
Comment (4tXJ7f) [inline src/api/java/cvc5/Solver.java:1082]:
```suggestion
   * Create a positive zero (+0.0) floating-point constant.
```

File: src/api/java/cvc5/Solver.java [hunk 3]
```diff
@@ -26,36 +27,2502 @@ public long getPointer()
...
+    Utils.validateUnsigned(sig, "sig");
+    long termPointer = mkPosZero(pointer, exp, sig);
+    return new Term(this, termPointer);
+  }
+
+  private native long mkPosZero(long pointer, int exp, int sig);
+
+  /**
+   * Create a negative zero (-0.0) floating-point constant. Requires CVC4 to
+   * be compiled with SymFPU support.
```
Comment (4tXJ7f) [inline src/api/java/cvc5/Solver.java:1099]:
```suggestion
   * Create a negative zero (-0.0) floating-point constant.
```

File: src/api/java/cvc5/Solver.java [hunk 4]
```diff
@@ -26,36 +27,2502 @@ public long getPointer()
...
+    Utils.validateUnsigned(index, "index");
+    long termPointer = mkAbstractValue(pointer, index);
+    return new Term(this, termPointer);
+  }
+
+  private native long mkAbstractValue(long pointer, long index);
+
+  /**
+   * Create a floating-point constant (requires CVC4 to be compiled with
+   * symFPU support).
```
Comment (4tXJ7f) [inline src/api/java/cvc5/Solver.java:1167]:
```suggestion
   * Create a floating-point constant
```

File: src/api/java/cvc5/Solver.java [hunk 5]
```diff
@@ -26,36 +27,2502 @@ public long getPointer()
...
+   * @param fun the sorted function
+   * @param bound_vars the parameters to this function
+   * @param term the function body
+   * @param global determines whether this definition is global (i.e. persists
+   *               when popping the context)
+   * @return the function
+   */
+  public Term defineFun(Term fun, Term[] bound_vars, Term term, boolean global)
+  {
+    long[] boundVarPointers = Utils.getPointers(bound_vars);
```
Comment (4tXJ7f) [inline src/api/java/cvc5/Solver.java:1608]:
```suggestion
  public Term defineFun(Term fun, Term[] boundVars, Term term, boolean global)
  {
    long[] boundVarPointers = Utils.getPointers(boundVars);
```

File: src/api/java/cvc5/Solver.java [hunk 6]
```diff
@@ -26,36 +27,2502 @@ public long getPointer()
...
+   * \endverbatim
+   * Create parameter 'fun' with mkConst().
+   * @param fun the sorted function
+   * @param bound_vars the parameters to this function
+   * @param term the function body
+   * @return the function
+   */
+  public Term defineFun(Term fun, Term[] bound_vars, Term term)
+  {
+    return defineFun(fun, bound_vars, term, false);
```
Comment (4tXJ7f) [inline src/api/java/cvc5/Solver.java:1590]:
```suggestion
  public Term defineFun(Term fun, Term[] boundVars, Term term)
  {
    return defineFun(fun, boundVars, term, false);
```

File: src/api/java/cvc5/Solver.java [hunk 7]
```diff
@@ -26,36 +27,2502 @@ public long getPointer()
...
+
+  /**
+   * Define n-ary function in the current context.
+   * SMT-LIB:
+   * \verbatim
+   * ( define-fun <function_def> )
+   * \endverbatim
+   * Create parameter 'fun' with mkConst().
+   * @param fun the sorted function
+   * @param bound_vars the parameters to this function
```
Comment (4tXJ7f) [inline src/api/java/cvc5/Solver.java:1584]:
```suggestion
   * @param boundVars the parameters to this function
```

File: src/api/java/cvc5/Solver.java [hunk 8]
```diff
@@ -26,36 +27,2502 @@ public long getPointer()
...
+  }
+  /**
+   * Define n-ary function.
+   * SMT-LIB:
+   * \verbatim
+   * ( define-fun <function_def> )
+   * \endverbatim
+   * Create parameter 'fun' with mkConst().
+   * @param fun the sorted function
+   * @param bound_vars the parameters to this function
```
Comment (4tXJ7f) [inline src/api/java/cvc5/Solver.java:1600]:
```suggestion
   * @param boundVars the parameters to this function
```

File: src/api/java/cvc5/Solver.java (hunk 7 above)
Comment (4tXJ7f) [reply src/api/java/cvc5/Solver.java:1584]: This applies to all instances of `bound_vars`.

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
Just some minor comments.

File: src/api/java/cvc5/Solver.java [hunk 9]
```diff
@@ -26,36 +27,2502 @@ public long getPointer()
...
+  public Term mkConstArray(Sort sort, Term val)
+  {
+    long termPointer = mkConstArray(pointer, sort.getPointer(), val.getPointer());
+    return new Term(this, termPointer);
+  }
+
+  private native long mkConstArray(long pointer, long sortPointer, long valPointer);
+  /**
+   * Create a positive infinity floating-point constant. Requires CVC4 to be
+   * compiled with SymFPU support.
```
[the remark this answers is not in the document]
Comment (mudathirmahgoub) [reply src/api/java/cvc5/Solver.java]: Done

File: src/api/java/cvc5/Solver.java [hunk 10]
```diff
@@ -26,36 +27,2502 @@ public long getPointer()
...
+    Utils.validateUnsigned(exp, "exp");
+    Utils.validateUnsigned(sig, "sig");
+    long termPointer = mkPosInf(pointer, exp, sig);
+    return new Term(this, termPointer);
+  }
+
+  private native long mkPosInf(long pointer, int exp, int sig);
+  /**
+   * Create a negative infinity floating-point constant. Requires CVC4 to be
+   * compiled with SymFPU support.
```
[the remark this answers is not in the document]
Comment (mudathirmahgoub) [reply src/api/java/cvc5/Solver.java]: Done

File: src/api/java/cvc5/Solver.java (hunk 1 above)
Comment (mudathirmahgoub) [reply src/api/java/cvc5/Solver.java]: Done

File: src/api/java/cvc5/Solver.java (hunk 3 above)
Comment (mudathirmahgoub) [reply src/api/java/cvc5/Solver.java]: Done

File: src/api/java/cvc5/Solver.java (hunk 4 above)
Comment (mudathirmahgoub) [reply src/api/java/cvc5/Solver.java]: Done

File: src/api/java/cvc5/Solver.java (hunk 2 above)
Comment (mudathirmahgoub) [reply src/api/java/cvc5/Solver.java]: Done

File: src/api/java/cvc5/Solver.java (hunk 5 above)
Comment (mudathirmahgoub) [reply src/api/java/cvc5/Solver.java]: Done

File: src/api/java/cvc5/Solver.java (hunk 6 above)
Comment (mudathirmahgoub) [reply src/api/java/cvc5/Solver.java]: Done

File: src/api/java/cvc5/Solver.java (hunk 7 above)
Comment (mudathirmahgoub) [reply src/api/java/cvc5/Solver.java]: Done

File: src/api/java/cvc5/Solver.java (hunk 8 above)
Comment (mudathirmahgoub) [reply src/api/java/cvc5/Solver.java]: Done

**Review by 4tXJ7f — APPROVED** [review]
LGTM, thank you for all the work!

**Pull request merged**